File: code-squad/week03/크레인-인형뽑기-게임.java
import java.util.Stack;
class Solution {
    public int solution(int[][] board, int[] moves) {
        int answer = 0;
        Stack<Integer> stack = new Stack<Integer>();

        for(int move : moves){
            int level = getLevel(board, move - 1);
            if(level >= 0){
                int target = board[level][move - 1];
                board[level][move - 1] = 0;
                stack.push(target);
                answer += updateStack(stack);
            }
        }
        return answer;
    }

    public int getLevel(int[][] board, int move){
        for(int i = 0; i < board.length; i++){
            if (board[i][move] > 0){
                return i;
            }
        }
        return -1;
    }

    public int updateStack(Stack<Integer> stack){
        int counter = 0;
        while(stack.size() > 1){
            Integer firstLast = stack.get(stack.size() - 1);
            Integer secondLast = stack.get(stack.size() - 2);
            if(Integer.compare(firstLast, secondLast) == 0){
                stack.pop();
                stack.pop();
                counter += 2;
            } else {
                break;
            }
        }
        return counter;

    }
}

File: code-squad/week03/크레인-인형뽑기-게임.py
from collections import deque
def solution(board, moves):
    answer = 0
    stack = deque([])
    for move in moves:
        get_level_from_board(move - 1, board, stack)
        answer += update_stack(stack)
    return answer

def get_level_from_board(move, board, stack):
    level = 0
    while level < len(board):
        target = board[level][move]
        if target != 0:
            stack.append(target)
            board[level][move] = 0
            break
        level += 1

def update_stack(stack):
    num_of_pop = 0
    while len(stack) > 1:
        if stack[-1] == stack[-2]:
            stack.pop()
            stack.pop()
            num_of_pop += 2
        else: break
    return num_of_pop

print(solution([[0, 0, 0, 0, 0], [0, 0, 1, 0, 3], [0, 2, 5, 0, 1], [4, 2, 4, 4, 2], [3, 5, 1, 3, 1]], [1, 5, 3, 5, 1, 2, 1, 4]))